refs/src/Effect/Ref.ss

```diff
@@ -6,38 +6,41 @@
           read
           modifyImpl
           write)
-  (import (prefix (chezscheme) scm:))
+  (import (prefix (chezscheme) scm:)
+          (prefix (purs runtime) rt:))
 
   (scm:define _new
     (scm:lambda (v)
       (scm:lambda ()
-        (scm:box v))))
+        (scm:list (scm:box v) (scm:make-mutex)))))
 
   (scm:define newWithSelf
     (scm:lambda (f)
       (scm:lambda ()
-        (scm:let ([ref (scm:box (scm:quote ()))])
-          (scm:set-box! ref (f ref))
+        (scm:let ([ref (scm:list (scm:box (scm:quote ())) (scm:make-mutex))])
+          (scm:set-box! (scm:car ref) (f ref))
           ref))))
 
   (scm:define read
     (scm:lambda (ref)
       (scm:lambda ()
-        (scm:unbox ref))))
+        (scm:unbox (scm:car ref)))))
 
   (scm:define modifyImpl
     (scm:lambda (f)
       (scm:lambda (ref)
         (scm:lambda ()
-          (scm:let* ([t (f (scm:unbox ref))]
-                     [v (scm:hashtable-ref t "state" (scm:quote Effect.Ref:modifyImpl-CANT-GET-STATE))])
-            (scm:set-box! ref v)
-            v)))))
+          (scm:with-mutex (scm:cadr ref)
```

Comment:
Should it be possible to use `box-cas!` instead? We would read the value from the box, then use `box-cas!` after that would only succeed if the value was still the same.

Reply:
Although, then if the swap fails, we'd need to try again.

Reply:
The documentation doesn't say if `with-mutex` does a blocking acquire. Do you have any insight on this?

Reply:
Sure. I'll just note the tradeoff that we'd have here: in high contention situations, CAS is a lot of busywork, mutexes just mostly sit around waiting.
That's probably fine for now, and in any case it's always possible to get a mutex-based primitive by just wrapping these in a mutex. We might want to offer both in the `stdlib` package.

Reply:
> The documentation doesn't say if `with-mutex` does a blocking acquire. Do you have any insight on this?

Mutexes for Chez are just bindings to the underlying POSIX stuff (i.e. `pthread_mutex_lock`), which has to acquire a piece of memory (or I don't see how it could possibly work otherwise).

One is also able to acquire memory manually, with `memory-order-acquire` (see section 15.7), and I suppose that's how Chez guarantees memory safety for the basic datastructures (since that works in a non-posix-threads environment too)

Reply:
What I mean is that `mutex-acquire` takes as argument whether the acquire should block or not, but `with-mutex` doesn't take this argument. I can't find from the source code what `with-mutex` does. I assume it calls `mutex-acquire` internally.

Reply:
How would the CAS-based implementation look like? Something like this?
```scheme
(scm:let* ([loop (lambda () 
                  (scm:let* ([old-val (scm:unbox ref)]
                             [new-val (f old-val)]
                             [v (rt:record-ref t (scm:quote state))]
                             [cas-result (scm:box-cas! ref old-val new-val)])
                    (cond [cas-result v] [else (loop)])))])
  (loop))
```

Reply:
CAS generally is. I can try to do a quick benchmark using the test in https://github.com/purescm/purescript-chez-stdlib/pull/1

Reply:
Well this is interesting - running the test above for the current Ref implementation and for a CAS implementation gives quite unexpected results:
```
Benching: high contention with mutex
mean   = 156.12 ms
stddev = 21.14 ms
min    = 115.09 ms
max    = 178.73 ms
Benching: high contention with CAS
mean   = 878.06 ms
stddev = 307.06 ms
min    = 5.68 ms
max    = 998.78 ms
```

The CAS implementation is basically what I posted above:
```
  (scm:define modifyImpl
    (scm:lambda (f)
      (scm:lambda (ref)
        (scm:lambda ()
          (scm:let loop ()
            (scm:let* ([old-val (scm:unbox ref)]
                       [result (f old-val)]
                       [new-val (rt:record-ref result (scm:quote state))]
                       [cas-result (scm:box-cas! ref old-val new-val)])
              (scm:cond [cas-result new-val]
                        [scm:else (loop)])))))))
```

Do we want to investigate why the CAS is so much slower or shall we just ship mutexes and defer the issue until later? This is an implementation detail and the external interface would not change anyways. I would keep the CAS implementation in the `stdlib` repo, so we can pick it up easily later.

Reply:
Using a mutex is fine.